encrypt/encrypt.go

```diff
@@ -27,14 +36,288 @@ var ArgHelp = `
+	encryptedFile, err := os.Open(fileSet.encrypted)
+	if err != nil {
+		return nil, err
+	}
+	defer encryptedFile.Close()
+
+	// encrypted md5
+	md5Hash.Reset()
+	shaHash.Reset()
+
+	tee = io.TeeReader(encryptedFile, md5Hash)
+	_, err = io.Copy(shaHash, tee)
+	if err != nil {
+		return nil, err
+	}
+	hashes.encryptedMd5 = hex.EncodeToString(md5Hash.Sum(nil))
+	hashes.encryptedSha256 = hex.EncodeToString(shaHash.Sum(nil))
+
+	return &hashes, nil
+}
+
+// Reads a public key file from a file using the crypt4gh keys module
+func readPublicKey(filename string) (key *[32]byte, err error) {
+	log.Info("Reading Public key file")
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	// This function panics if the key is malformed, so we handle that as well
+	// as errors
+	defer func() {
+		log.Info("Hey! this is a panic!")
+		if recover() != nil {
+			err = fmt.Errorf("malformed key file: %s", filename)
+		}
+	}()
+
+	publicKey, err := keys.ReadPublicKey(file)
+
+	return &publicKey, err
+}
+
+// Generates a crypt4gh key pair, returning only the private key, as the
+// public key used for encryption is the one provided as argument.
+func generatePrivateKey() (*[32]byte, error) {
+	log.Info("Generating encryption key")
+
+	_, privateKey, err := keys.GenerateKeyPair()
+	if err != nil {
+		return nil, errors.New("failed to generate private key for encryption")
+	}
+	return &privateKey, nil
+}
+
+// Encrypts the data from `filename` into `outFilename` for the given `pubKey`,
+// using the given `privateKey`.
+func encrypt(filename, outFilename string, pubKey, privateKey [32]byte) error {
+	// check if outfile exists
+	if FileExists(outFilename) {
+		return fmt.Errorf("outfile %s already exists", outFilename)
+	}
+
+	// read infile
+	inFile, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	defer inFile.Close()
+
+	// open outfile for writing
+	outFile, err := os.Create(outFilename)
+	if err != nil {
+		return err
+	}
+	defer outFile.Close()
+
+	// Create crypt4gh writer
+	pubKeyList := [][32]byte{pubKey}
+	crypt4GHWriter, err := streaming.NewCrypt4GHWriter(outFile,
+		privateKey, pubKeyList, nil)
+	if err != nil {
+		return err
+	}
+	defer crypt4GHWriter.Close()
+
+	// Encrypt the data
+	_, err = io.Copy(crypt4GHWriter, inFile)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+//
+// Helper functions and structs
+//
+
+// struct type to keep track of infiles and outfiles for encryption and
+// decryption
+type encryptionFileSet struct {
+	unencrypted string
+	encrypted   string
+}
+
+// struct to keep track of all the checksums for a given unencrypted input file.
+type hashSet struct {
+	encryptedMd5      string
+	unencryptedMd5    string
+	encryptedSha256   string
+	unencryptedSha256 string
+}
+
+// Checks if a file exists in the file system. Note that this function will not
+// check if the file is readable, or if the file is a directory, only if it
+// exists.
+func FileExists(filename string) bool {
```

Comment:
```suggestion
func fileExists(filename string) bool {
```

Function names starting with a capital letter gets exported, no need since it is only used in this module

Reply:
Yes - but I wanna use it in the decrypt module as well, so I'm exporting it for that!

Reply:
Isn't it better to put shared functions directly in main?